README.md
# spectral-format
Parsers for various file formats


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.


## [0.3.1] - 2020-12-25
### Added:
- WAVE chunk handlers (fmt/data)
- RIFF data printout
- Debug memory hook submodule


## [0.1.4] - 2020-12-25
### Added:
- 'format_t' class
- 'format::riff' class
- Default RIFF chunk handler (copy)



test/test_format_riff.cpp
#include <cstdio>

#include <iostream>
#include <fstream>

#include <CppUnitTest.h>

#include "format.h"
#include "format_riff.h"
#include "format_wav.h"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace spectral {
namespace test {


TEST_CLASS(test_format_riff) {
public:
  bool TestCopy( format::riff &File ) {
    std::string 
      FileNameIn = SolutionDir "assets/wav_1m.wav",
      FileNameOut = SolutionDir "out/wav_1m.wav";

    File.Read(FileNameIn);
    File.Write(FileNameOut);

    FILE
      *FileIn = fopen(FileNameIn.c_str(), "rb"),
      *FileOut = fopen(FileNameOut.c_str(), "rb");

    if (FileOut == nullptr) {
      fclose(FileOut);
      return false;
    }
    fseek(FileIn, 0, SEEK_END);
    int FInSize = ftell(FileIn);
    fseek(FileIn, 0, SEEK_SET);
    fseek(FileOut, 0, SEEK_END);
    int FOutSize = ftell(FileOut);
    fseek(FileOut, 0, SEEK_SET);

    if (FInSize != FOutSize) {
      fclose(FileOut);
      fclose(FileIn);
      return false;
    }

    format::byte
      *BufIn = new format::byte[FInSize],
      *BufOut = new format::byte[FInSize];
    fread(BufIn, sizeof(format::byte) * FInSize, 1, FileIn);
    fread(BufOut, sizeof(format::byte) * FInSize, 1, FileOut);
    for (int i = 0; i < FInSize; ++i)
      if (BufIn[i] != BufOut[i]) {
        fclose(FileOut);
        fclose(FileIn);
        return false;
      }

    fclose(FileOut);
    fclose(FileIn);
    return true;
  }

  TEST_METHOD(TestRWDefault) {
    format::riff File;
    Assert::IsTrue(TestCopy(File));
  }


  TEST_METHOD(TestRW_RIFF_WAVE) {
    format::riff File;
    File
      << std::make_pair<>(format::chunk::wav_format::Tag, std::make_shared<format::chunk::wav_format::factory>())
      << std::make_pair<>(format::chunk::wav_data::Tag, std::make_shared<format::chunk::wav_data::factory>());
    Assert::IsTrue(TestCopy(File));
  }


  TEST_METHOD(Test_WAVE_Output) {
    format::riff File;
    File
      << std::make_pair<>(format::chunk::wav_format::Tag, std::make_shared<format::chunk::wav_format::factory>())
      << std::make_pair<>(format::chunk::wav_data::Tag, std::make_shared<format::chunk::wav_data::factory>());
    File.Read(SolutionDir "assets/wav_1m.wav");
    std::ofstream Out(SolutionDir "out/TEST_WAVE_Output.out");
    File.Print(Out);
  }
};


} /* End of 'test' namespace */
} /* End of 'spectral' namespace */


src/format_wav.h
/***************************************************************
 * Copyright (C) 2020
 *    Reshetnyak Ivan
 ***************************************************************/

#pragma once

#include <stdint.h>

#include "format_riff.h"

namespace spectral {
namespace format {
namespace chunk {


class wav_format : public riff::chunk_t {
public:
  static const dword Tag = ' ' << 24 | 't' << 16 | 'm' << 8 | 'f';  // "fmt "
  word Format, NumOfChannels, BlockAlign;
  dword SamplingRate, AverageBPS;
  std::unique_ptr<byte[]> SpecificFields;

  class factory : public chunk_t::factory {
  public:
    factory() = default;
    virtual ~factory() = default;
    virtual std::shared_ptr<chunk_t> Create( dword ID, dword Size ) const override;
  };
  friend factory;

  virtual ~wav_format() = default;
  virtual void Clear() override;
  virtual void Write( FILE *File ) const override;
  std::ostream & Print( std::ostream &Stream ) const override;
  virtual void Read( FILE *File ) override;

protected:
  static const int CommonFieldSize = sizeof(word) * 3 + sizeof(dword) * 2;
  wav_format( dword ID, dword Size );
};


class wav_data : public riff::chunk_t {
public:
  static const int32_t Tag = 'a' << 24 | 't' << 16 | 'a' << 8 | 'd';  // "data"
  std::unique_ptr<byte[]> Data;

  class factory : public chunk_t::factory {
  public:
    factory() = default;
    virtual ~factory() = default;
    virtual std::shared_ptr<chunk_t> Create( dword ID, dword Size ) const override;
  };
  friend factory;

  virtual ~wav_data() = default;
  virtual void Clear() override;
  virtual void Write( FILE *File ) const override;
  std::ostream & Print( std::ostream &Stream ) const override;
  virtual void Read( FILE *File ) override;

protected:
  wav_data( dword ID, dword Size );
};


} /* End of 'chunk' namespace */
} /* End of 'format' namespace */
} /* End of 'spectral' namespace */


src/format_riff.h
/***************************************************************
 * Copyright (C) 2020
 *    Reshetnyak Ivan
 ***************************************************************/

#pragma once

#include <cstdio>
#include <stdint.h>

#include <iostream>
#include <memory>
#include <unordered_map>
#include <vector>

#include "format.h"

namespace spectral {
namespace format {


class riff : public format_t {
public:
  using buffer = std::shared_ptr<byte[]>;


  class chunk_t {
  public:
    dword ID, Size;

    class factory {
    public:
      factory() = default;
      virtual ~factory() = default;
      virtual std::shared_ptr<chunk_t> Create( dword ID, dword Size ) const = 0;
    };
    friend factory;

    virtual ~chunk_t() = default;
    virtual void Clear();
    virtual void Write( FILE *File ) const = 0;
    virtual void Read( FILE *File ) = 0;
    virtual std::ostream & Print( std::ostream &Stream ) const;
    void SkipPadByte( FILE *File ) const;
    void AddPadByte( FILE *File ) const;

  protected:
    chunk_t( dword ID, dword Size );
  };

protected:
  using mapping = std::unordered_map<dword, std::shared_ptr<chunk_t::factory>>;

public:
  riff();
  riff( std::shared_ptr<chunk_t::factory> DefaultFactory );
  virtual ~riff() = default;
  virtual void Read( const std::string &FileName ) override;
  virtual void Write( const std::string &FileName ) const override;
  virtual std::ostream & Print( std::ostream &Stream ) const;
  riff & operator<<( const mapping::value_type &CustomChunk );
  std::shared_ptr<chunk_t> Produce( dword ID, dword Size ) const;
  std::vector<std::shared_ptr<chunk_t>> Chunks;

protected:
  mapping CustomChunks;
  std::shared_ptr<chunk_t::factory> DefaultFactory;
};


namespace chunk {


class unknown final : public riff::chunk_t {
public:
  class factory final : public chunk_t::factory {
  public:
    factory() = default;
    ~factory() = default;
    std::shared_ptr<chunk_t> Create( dword ID, dword Size ) const override;
  };
  friend factory;

  ~unknown() = default;
  void Clear() final override;
  void Write( FILE *File ) const final override;
  void Read( FILE *File ) final override;

private:
  unknown( dword ID, dword Size );
  std::unique_ptr<byte[]> Data;
};


class riff : public ::spectral::format::riff::chunk_t {
public:
  static const dword Tag = 'F' << 24 | 'F' << 16 | 'I' << 8 | 'R';  // RIFF
  dword Format;
  std::vector<std::shared_ptr<chunk_t>> Subchunks;

  class factory : public chunk_t::factory {
  public:
    factory( const ::spectral::format::riff &Mapping );
    virtual ~factory() = default;
    virtual std::shared_ptr<chunk_t> Create( dword ID, dword Size ) const override;

  protected:
    const ::spectral::format::riff &Mapping;
  };
  friend factory;

  virtual ~riff() = default;
  virtual void Clear() override;
  virtual void Read( FILE *File ) override;
  virtual void Write( FILE *File ) const override;
  std::ostream & Print( std::ostream &Stream ) const override;

protected:
  const ::spectral::format::riff &Mapping;
  riff( const ::spectral::format::riff &Mapping, dword ID, dword Size );
};


} /* End of 'chunk' namespace */
} /* End of 'format' namespace */
} /* End of 'spectral' namespace */


src/format_wav.cpp
/***************************************************************
 * Copyright (C) 2020
 *    Reshetnyak Ivan
 ***************************************************************/

#include "pch.h"

#include "format_wav.h"

namespace spectral {
namespace format {
namespace chunk {


std::shared_ptr<::spectral::format::riff::chunk_t> wav_format::factory::Create( dword ID, dword Size ) const {
  return std::shared_ptr<wav_format>(new wav_format(ID, Size));
}


void wav_format::Clear() {
  chunk_t::Clear();
  SpecificFields = nullptr;
}


void wav_format::Write( FILE *File ) const {
  fwrite(&ID, sizeof(ID), 1, File);
  fwrite(&Size, sizeof(Size), 1, File);
  int BytesWritten = 0;
  BytesWritten += fwrite(&Format, sizeof(Format), 1, File) * sizeof(Format);
  BytesWritten += fwrite(&NumOfChannels, sizeof(NumOfChannels), 1, File) * sizeof(NumOfChannels);
  BytesWritten += fwrite(&SamplingRate, sizeof(SamplingRate), 1, File) * sizeof(SamplingRate);
  BytesWritten += fwrite(&AverageBPS, sizeof(AverageBPS), 1, File) * sizeof(AverageBPS);
  BytesWritten += fwrite(&BlockAlign, sizeof(BlockAlign), 1, File) * sizeof(BlockAlign);
  if (BytesWritten < Size)
    fwrite(&SpecificFields[0], Size - BytesWritten, 1, File);
  AddPadByte(File);
}


std::ostream & wav_format::Print( std::ostream &Stream ) const {
  return Stream <<
    "fmt chunk with size: " << Size << " bytes begin" << std::endl <<
    "Format: " << Format << "; Number of channels: " << NumOfChannels <<
    "; Block alignment: " << BlockAlign << "; Sample rate: " << SamplingRate <<
    "; Average BPS:" << AverageBPS << "; " << std::endl <<
    Size - CommonFieldSize << " bytes of format-specific fields" << std::endl <<
    "fmt chunk end" << std::endl;
}


void wav_format::Read( FILE *File ) {
  int BytesRead = 0;
  BytesRead += fread(&Format, sizeof(Format), 1, File) * sizeof(Format);
  BytesRead += fread(&NumOfChannels, sizeof(Format), 1, File) * sizeof(NumOfChannels);
  BytesRead += fread(&SamplingRate, sizeof(SamplingRate), 1, File) * sizeof(SamplingRate);
  BytesRead += fread(&AverageBPS, sizeof(AverageBPS), 1, File) * sizeof(AverageBPS);
  BytesRead += fread(&BlockAlign, sizeof(BlockAlign), 1, File) * sizeof(BlockAlign);
  if (BytesRead < Size) {
    SpecificFields = std::make_unique<byte[]>(Size - BytesRead);
    fread(&SpecificFields[0], Size - BytesRead, 1, File);
  }
  SkipPadByte(File);
}


wav_format::wav_format( dword ID, dword Size ) : chunk_t(ID, Size) {
}


std::shared_ptr<::spectral::format::riff::chunk_t> wav_data::factory::Create( dword ID, dword Size ) const {
  return std::shared_ptr<wav_data>(new wav_data(ID, Size));
}


void wav_data::Clear() {
  chunk_t::Clear();
  Data = nullptr;
}


void wav_data::Write( FILE *File ) const {
  fwrite(&ID, sizeof(ID), 1, File);
  fwrite(&Size, sizeof(Size), 1, File);
  if (Data != nullptr)
    fwrite(&Data[0], Size, 1, File);
  AddPadByte(File);
}


std::ostream & wav_data::Print( std::ostream &Stream ) const {
  return Stream << "WAVE chunk with size: " << Size << " bytes" << std::endl;
}


void wav_data::Read( FILE *File ) {
  Data = std::make_unique<byte[]>(Size);
  fread(&Data[0], Size, 1, File);
  AddPadByte(File);
}


wav_data::wav_data( dword ID, dword Size ) : chunk_t(ID, Size) {
}


} /* End of 'chunk' namespace */
} /* End of 'format' namespace */
} /* End of 'spectral' namespace */


src/format_riff.cpp
/***************************************************************
 * Copyright (C) 2020
 *    Reshetnyak Ivan
 ***************************************************************/

#include "pch.h"

#include "format_riff.h"

namespace spectral {
namespace format {


riff::chunk_t::chunk_t( dword ID, dword Size ) : ID(ID), Size(Size) {
}


void riff::chunk_t::Clear() {
  ID = (dword)0;
  Size = (dword)0;
}


void riff::chunk_t::SkipPadByte( FILE *File ) const {
  if (Size % 2 == 1)
    fseek(File, 1, SEEK_CUR);
}


void riff::chunk_t::AddPadByte( FILE *File ) const {
  if (Size % 2 == 1) {
    byte Pad;
    fwrite(&Pad, sizeof(Pad), 1, File);
  }
}


std::ostream & riff::chunk_t::Print( std::ostream &Stream ) const {
  return Stream <<
    "Unrecognised chunk with ID: " << ID << "(" << (char)(ID & 0xFF) <<
    (char)(ID >> 8 & 0xFF) << (char)(ID >> 16 & 0xFF) << (char)(ID >> 24 & 0xFF) <<
    "), size: " << Size << " bytes" << std::endl;
}


riff::riff() : DefaultFactory(std::make_shared<chunk::unknown::factory>()) {
  this->operator<<(std::make_pair<>(chunk::riff::Tag, std::make_shared<chunk::riff::factory>(*this)));
}


riff::riff( std::shared_ptr<chunk_t::factory> DefaultFactory ) : DefaultFactory(DefaultFactory) {
}


void riff::Read( const std::string &FileName ) {
  FILE *File = fopen(FileName.c_str(), "rb");
  if (File == nullptr)
    return;

  fseek(File, 0, SEEK_END);
  int FileSize = ftell(File);
  fseek(File, 0, SEEK_SET);
  while (ftell(File) < FileSize) {
    dword ID, Size;
    fread(&ID, sizeof(ID), 1, File);
    fread(&Size, sizeof(Size), 1, File);

    auto NewChunk = Produce(ID, Size);
    NewChunk->Read(File);
    Chunks.push_back(NewChunk);
  }
  fclose(File);
}


void riff::Write( const std::string &FileName ) const {
  FILE *File = fopen(FileName.c_str(), "wb");
  if (File == nullptr)
    return;

  for (const auto &Chunk : Chunks)
    Chunk->Write(File);
  fclose(File);
}


std::ostream & riff::Print( std::ostream &Stream ) const {
  Stream << "RIFF file begin" << std::endl;
  for (const auto &Chunk : Chunks)
    Chunk->Print(Stream);
  Stream << "RIFF file end" << std::endl;
  return Stream;
}


riff & riff::operator<<( const mapping::value_type &CustomChunk ) {
  CustomChunks.insert(CustomChunk);
  return *this;
}


std::shared_ptr<riff::chunk_t> riff::Produce( dword ID, dword Size ) const {
  auto Factory = CustomChunks.find(ID);
  if (Factory != CustomChunks.end())
    return Factory->second->Create(ID, Size);
  return DefaultFactory->Create(ID, Size);
}


namespace chunk {


std::shared_ptr<::spectral::format::riff::chunk_t> unknown::factory::Create( format::dword ID, format::dword Size ) const {
  return std::shared_ptr<unknown>(new unknown(ID, Size));
}


void unknown::Clear() {
  chunk_t::Clear();
  Data = nullptr;
}


void unknown::Write( FILE *File ) const {
  fwrite(&ID, sizeof(ID), 1, File);
  fwrite(&Size, sizeof(Size), 1, File);
  fwrite(&Data[0], sizeof(Data[0]) * Size, 1, File);
  AddPadByte(File);
}


void unknown::Read( FILE *File ) {
  Data = std::make_unique<byte[]>(Size);
  fread(&Data[0], sizeof(Data[0]) * Size, 1, File);
  SkipPadByte(File);
}


unknown::unknown( dword ID, dword Size ) : chunk_t(ID, Size) {
}


riff::factory::factory( const ::spectral::format::riff &Mapping ) : Mapping(Mapping) {
}


std::shared_ptr<::spectral::format::riff::chunk_t> riff::factory::Create( dword ID, dword Size ) const {
  return std::shared_ptr<riff>(new riff(Mapping, ID, Size));
}


void riff::Clear() {
  chunk_t::Clear();
  Subchunks.clear();
}


void riff::Write( FILE *File ) const {
  fwrite(&ID, sizeof(ID), 1, File);
  fwrite(&Size, sizeof(Size), 1, File);
  fwrite(&Format, sizeof(Format), 1, File);
  for (const auto &Subchunk : Subchunks)
    Subchunk->Write(File);
  AddPadByte(File);
}


std::ostream & riff::Print( std::ostream &Stream ) const {
  Stream << "RIFF chunk begin, size: " << Size << " bytes" << std::endl;
  for (const auto &Subchunk : Subchunks)
    Subchunk->Print(Stream);
  Stream << "RIFF chunk end" << std::endl;
  return Stream;
}


void riff::Read( FILE *File ) {
  fread(&Format, sizeof(Format), 1, File);

  int BytesRead = sizeof(Format);
  while (BytesRead < Size) {
    dword ID, Size;
    fread(&ID, sizeof(ID), 1, File);
    fread(&Size, sizeof(Size), 1, File);

    auto NewSubchunk = Mapping.Produce(ID, Size);
    NewSubchunk->Read(File);
    Subchunks.push_back(NewSubchunk);
    BytesRead += sizeof(ID) + sizeof(Size) + Size + Size % 2;
  }
  SkipPadByte(File);
}


riff::riff( const ::spectral::format::riff &Mapping, dword ID, dword Size ) :
    chunk_t(ID, Size), Mapping(Mapping) {
}


} /* End of 'chunk' namespace */
} /* End of 'format' namespace */
} /* End of 'spectral' namespace */
